src/utils.c

```diff
@@ -281,12 +281,42 @@ ts_timestamp_bucket(PG_FUNCTION_ARGS)
 {
 	Interval   *interval = PG_GETARG_INTERVAL_P(0);
 	Timestamp	timestamp = PG_GETARG_TIMESTAMP(1);
+	Timestamp	origin;
 	Timestamp	result;
 	int64		period = -1;
 
+	if (PG_NARGS() > 2)
+		origin = PG_GETARG_TIMESTAMPTZ(2);
+	else
+	{
+		/*
+		 * The default origin moves the PG epoch to start on a monday:
+		 * 2000-01-03. We don't use PG epoch since it starts on a saturday.
+		 * This makes time-buckets by a week more intuitive and aligns it with
+		 * date_trunc.
+		 */
+#ifdef HAVE_INT64_TIMESTAMP
+		origin = 2 * USECS_PER_DAY;
+#else
+		origin = 2 * SECS_PER_DAY;
+#endif
+	}
+
 	if (TIMESTAMP_NOT_FINITE(timestamp))
 		PG_RETURN_TIMESTAMP(timestamp);
 
+	if (origin > 0 && timestamp < DT_NOBEGIN + origin)
```

Comment:
Isn't it better to calculate the remainder of the origin and use that as the shift for the result (e.g., `shift = origin % period`)? That will allow much larger number of origin values without overflowing the timestamp and you would avoid having to first subtract the origin value to compensate. 

Or, am I missing something with how this works?

Reply:
In fact you are right the shift can be  `shift = origin % period` That's fixed. However the shift still needs to be subtracted before truncation and added back after.